### DISCUSSION TITLE
A new subsystem for children nodes positioning in layouts

### DISCUSSION THREAD

**Discussion by sifran-github** [freeplane/freeplane#1101] [General]
After thinking about it for a long time, I present to you a new way to describe the properties of layouts, that will allow more clarity, control and options when creating new layouts.

# The Problem
In my opinion, the current layouts properties are not enough to represent an exact layout.

For example, as i wrote in another thread, what exactly this icon represent?
![image](https://user-images.githubusercontent.com/5730355/232353124-02519a3f-6f69-41a8-acd0-c9f60fd6319d.png)

- does it represent that the first child's horizontal position start from the left side of the parent, with some gap?
- does it represent that the first child's horizontal position start closer to the right side of the parent?
- does it represent that the first child's horizontal position start from the center of the parent?

Today, the answer is not clear, and it creates some issues.
if we assume the first point is correct, then we might get some unintended results when the parent is very big compared to his children:

![image](https://user-images.githubusercontent.com/5730355/232353497-95a0182f-d598-4bb5-ac0c-6dbfbb77bdf7.png)

And although in both cases the direction is to the RIGHT, it looks like the nodes are staying over the left of the parent. this is why the thread https://github.com/freeplane/freeplane/discussions/1095 was opened.

BUT if we just choose another behavior of the same icon, then in other times we will get another unintended result. and we dont have control over it.

This is just one example of layout that doesn't fit the expected result that the icon represent, but there are more.

And here my suggested solution.

# The Solution

Lets take the parent and the child. each one of them has 3 "important" positions horizontally (left,center,right), and 3 vertically (top,center,bottom). together each has 9 points:
![image](https://user-images.githubusercontent.com/5730355/232353736-c025ca79-ef86-42b6-9298-7e05ce17d0df.png)

Maybe we can call them ANCHORS but i will refer to them as POSITION.

Lets now define a position relationship between them:
Parent's (LEFT, TOP) is connected to child (LEFT,BOTTOM), and it looks like this:
![image](https://user-images.githubusercontent.com/5730355/232353971-f3c3958a-f52e-4cc4-b842-5a12b45cdcbe.png)

Nice! and now we just need to add CHILD_MARGIN (from this point, move the child in a specific direction, to create "gap" between the parent and the child. has 2 values, horizontal and vertical), CHILDREN_DIRECTION (UP,DOWN,LEFT,RIGHT) and CHILDREN_GAP (gap beween one child to another child. just one value).

So lets add the properties:
- PARENT_POSITION = (LEFT, TOP)
- CHILD_POSITION = (LEFT, BOTTOM)
- CHILD_MARGIN = (+100, -50)
- CHILDREN_DIRECTION = RIGHT
- CHILDREN_GAP = 100.

and we will get this:
![image](https://user-images.githubusercontent.com/5730355/232354647-e8d74328-abe9-4b2e-8987-9ba8de3d512f.png)

and because we defined it to be connected to the left, this will be our result if the parent is large:

![image](https://user-images.githubusercontent.com/5730355/232354716-cf7cd098-2bac-46ac-a433-de66394c583b.png)

but if we want to change it and make sure that the children are goint to stay on the right side of the parent, lets just "stick them" to the right side by changing the properties to:

- PARENT_POSITION = (RIGHT,TOP) <--- NOTICE THE "RIGHT" POSITION.
- CHILD_POSITION = (LEFT, BOTTOM)
- CHILD_MARGIN = (-100, -50)
- CHILDREN_DIRECTION = RIGHT
- CHILDREN_GAP = 100.

and now we have this result:

![image](https://user-images.githubusercontent.com/5730355/232355002-8d1288e0-ed14-4d08-9325-6d9f5ce348ca.png)

as you can see, it is a simple subsystem of properties that allows the configuration of many layout to be defined.

# GUI
How to represent it? do we allow the user to define all of these parameters by himself? wouldn't it be too much to configure?

So here I have a suggestion but maybe someone can improve on it. my suggestion is to let the user configure every parameter by himself, and leave the layout icons as a "template" that will only load specific parameters of that specific layout, and the user can continue to change the parameters from there.

You want a specific "outline view"? configure it by yourself or click the layout icon to load the "layout template" and edit some values from there.

This will allow both configurability of everything, and ease of use.

# NOTES

- I took some of the terminology from CSS web design. They struggle the issue of positioning elements for years.
- There probably be another property CHILDREN_ON_BOTH_SIDES with values TRUE or FALSE.
- I think that with these parameters, the "base direction to children" parameter can be removed.
- Maybe there is a need for another parameter or two, of edges locations.
- Instead of 3 values for horizontal (LEFT, CENTER, RIGHT) we can use percentage (0%-100%). meaning it will be percentage of the length of the node. same for vertical. But: anyway it will not replace margin which is constant, and I think that its an overkill. these 3 values, combined with margin, are enough to create all the meaningful layouts, in my opinion.

If someone wants the PowerPoint file used to create the images here: 
[new_layout_children_design.zip](https://github.com/freeplane/freeplane/files/11245901/new_layout_children_design.zip)

**Comment by sifran-github** [comment]
## MORE EXAMPLES

![image](https://user-images.githubusercontent.com/5730355/232362796-7c3564d1-713b-41ff-97d8-eb9211519dcb.png)

**_EXAMPLE 1_**
PARENT_POSITION = (LEFT, BOTTOM)
CHILD_POSITION = (LEFT, TOP)
CHILD_MARGIN = (100, 50)
CHILDREN_DIRECTION = DOWN
CHILDREN_GAP = 50.

**_EXAMPLE 2_**
PARENT_POSITION = (RIGHT, BOTTOM)
CHILD_POSITION = (LEFT, TOP)
CHILD_MARGIN = (100, 50)
CHILDREN_DIRECTION = DOWN
CHILDREN_GAP = 50.

**_EXAMPLE 3_**
PARENT_POSITION = (RIGHT, BOTTOM)
CHILD_POSITION = (TOP, RIGHT)
CHILD_MARGIN = (0, 50)
CHILDREN_DIRECTION = DOWN
CHILDREN_GAP = 50.

**_EXAMPLE 4_**
PARENT_POSITION = (BOTTOM,CENTER)
CHILD_POSITION = (LEFT, TOP)
CHILD_MARGIN = (50, 50)
CHILDREN_DIRECTION = DOWN
CHILDREN_GAP = 50.
CHILDREN_SIDES = 2.

## NOTE ABOUT TWO SIDES (Example 4)
this is one way of implementing 2 sides.
Freeplane can understand that because (CHILDREN_SIDES == 2) and (CHILDREN_DIRECTION == DOWN), it means that child positions will be LEFT AND RIGHT, even if the definition is CHILD_POSITION is defined as (LEFT, TOP), and it will take into account the first part of the margin (the horizontal one) and implement it on both sides.

**Comment by sifran-github** [comment]
## EXAMPLES with different CHILDREN_DIRECTION

![image](https://user-images.githubusercontent.com/5730355/232364933-e6e0667e-6d60-4ee8-82c6-e5aa5fb61e4e.png)

**_EXAMPLE 1R_**
Changed CHILDREN_DIRECTION to RIGHT.

**_EXAMPLE 2R_**
Changed CHILDREN_DIRECTION to RIGHT.

**_EXAMPLE 3L_**
Changed CHILDREN_DIRECTION to LEFT.

**_EXAMPLE 4R_**
Changed CHILDREN_DIRECTION to RIGHT.
(CHILDREN_SIDES is still 2).

**Comment by sifran-github** [comment]
## NOTE ABOUT CHILDREN ALIGNMENT.
In all of the examples, we assume the child alignment between themselves from the CHILD_POSITION + CHILDREN_DIRECTION  variables.

## Example 3:
![image](https://user-images.githubusercontent.com/5730355/232365669-03c2f281-a702-4687-8665-76eff52bb58c.png)
you can see the alignment was assumed automatically.

if we want to customize it even more to create something like this, where the alignment side is different:
![image](https://user-images.githubusercontent.com/5730355/232365838-a501d158-a4d5-4cef-878a-1f2833df86da.png)

we will need another parameter, CHILDREN_ALIGNMENT_POSITION, that we can set to a different value, in this case LEFT.

I think that... maybe we can skip these type of layouts. It does allow even more customization, but even without this custom alignment parameter, we can create a similar layout by aligning the left side to the center of the parent, for example:
![image](https://user-images.githubusercontent.com/5730355/232363798-1f534a42-6914-4f57-ab8f-4331502512f8.png)
Although there is a slight difference between these layouts, specially when we resize it.

## ANOTHER EXAMPLE:
![image](https://user-images.githubusercontent.com/5730355/232368194-0293d437-c91f-4495-9aa6-c108693dcba2.png)
In all of these configurations:
PARENT_POSITION = (RIGHT, CENTER)
CHILD_MARGIN = (100, 0)
CHILDREN_DIRECTION = RIGHT
CHILDREN_GAP = 50.

The only difference is where do we attach the child:
5-A: CHILD_POSITION = (LEFT, TOP)
5-B: CHILD_POSITION = (LEFT, CENTER)
5-C: CHILD_POSITION = (LEFT, BOTTOM)

And how do we calculate the correct alignment? Because CHILDREN_DIRECTION==RIGHT, it means that we look at the VERTICAL position of the CHILD_POSITION parameter and decide where to align all the other children.

**Comment by sifran-github** [comment]
## CONNECTION TO THE BOX INSTEAD OF A SPECIFIC CHILD.
I also have to address these special layouts:
![image](https://user-images.githubusercontent.com/5730355/232453071-a1a5f9ac-d022-4bd8-9add-4517ba407d8c.png)

These layouts do not connect to a specific child, but rather they connect to the BOX OF THE FULL TREE that all of the children are creating.

for example:

![image](https://user-images.githubusercontent.com/5730355/232454091-49f16832-ee0e-4897-b29e-384b52ba7ad2.png)

The orange box is the full tree box. the properties between the root and his children are:
PARENT_POSITION = (RIGHT, CENTER)
CHILD_POSITION = (BOX_LEFT, BOX_CENTER) <--- Notice the change here.
CHILD_MARGIN = (100, 0)
CHILDREN_DIRECTION = DOWN
CHILDREN_GAP = 50.
CHILDREN_ON_BOTH_SIDES = FALSE

If we want both sides:
![image](https://user-images.githubusercontent.com/5730355/232456028-4c1bb348-4deb-4eb4-8272-628ea4527a31.png)

The settings are the same, but the only change is:
CHILDREN_ON_BOTH_SIDES = TRUE

Note that although examples 6+7 might sound complicated, they the exact default layout of Freeplane since forever:

![image](https://user-images.githubusercontent.com/5730355/232457497-08ba3b50-6420-4032-bb77-95512ac6a565.png)

We can create even many more options other than connecting to the center of the full box, like "connect to 25% of the box height", but I think that its an overkill.

**Comment by sifran-github** [comment]
no comments? It took me a long time to think, come up with this new idea and publish it.

Even if this suggestion is a bit complicated and maybe suitable for future releases, I still think that its a very good idea, that will bring:
1. More control over layouts by providing new set of properties for layouts. And more clarity on how layouts should appear on screen.
2. Provide more configurability by the user: Instead of just picking a specific layout, you will be able to configure its properties.

**Reply by dpolivaev** [reply]
I want to switch from improving layouts to some other topics now.
Personally, I still think that different layout options are edge use cases not relevant for most Freeplane users.

**Reply by ClaytonCurtis1947** [reply]
I don't know what the underpinnings are of the current layout mechanism, but I like the author's idea as a unifying framework for layouts. One might avoid complicating the UI unduly by providing an "Advanced" or "User-defined layouts" button that could provide access to the approach the author describes.  I know the author put a lot of thought -- as well as time and effort -- into into this, and kudos to him.  In the interests of full transparency, I'm relatively new to FreePlane, and I have problems interpreting the iconic depictions (and their tool tips) of layouts as they are! :)

**Reply by dpolivaev** [reply]
Freeplane is a one-man show unless the community starts contributing. 
Generally, all layouts with a gear symbol are converted to some layout without a gear symbol depending on the layout of their parents.
How exactly it works can only be learned if you test it.

If you want to make a suggestion on how to change icons or descriptions, make concrete suggestions and I will check them and use them if I like them. 

If you want to introduce another UI send me a mockup. You currently can use the online mockup tool https://moqups.com/ for free with no significant limitations.

**Comment by ClaytonCurtis1947** [comment]
Having been in the same position (i.e., one-man author / maintainer of an application). I devoutly respect the difficult of the task and cannot express my gratitude for your work.

I completely understand your statement that “Freeplane is a one-man show unless the community starts contributing” and the frustration that must go with it.

Is there any way I can contribute / help?

Clayton Curtis

**Reply by dpolivaev** [reply]
Sure. But I do not know what exactly you could contribute, it is up to you.

**Comment by sifran-github** [comment]
> like the author's idea as a unifying framework for layouts.
>I know the author put a lot of thought -- as well as time and effort -- into into this, and kudos to him.

@ClaytonCurtis1947 thank you for your kind words.

@dpolivaev this thread and the contributions I make in the discussion forums are all about me trying to help the community.
Although I am a programmer, java is not my strong side, and the codebase is a bit too complicated for me. This is why I stay on the design/product level: to suggest ideas and help users.

I understand that implementing the idea in this thread requires more work, and maybe you don't want to do it after all the effort invested into the existing layouts, and I understand that and appreciate the work that has been done.

I wanted to write down this idea in order to "make sense" for myself and describe a more precise/simple/unified system for representing layouts. I hope you take it positively and also appreciate the effort.
I will continue to contribute in whatever I can.
